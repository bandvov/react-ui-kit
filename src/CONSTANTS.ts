export const COLORS = {
  Blue: "#02489F",
  green: "#00A046",
  red: "#E50000",
  orange: "#E57B00",
  gray: "#ADB3BF",
  lightgrey: "rgba(173, 179, 191, 0.5)"
}
export const STATUS = {
  success: "success",
  successOutlined: "success-outlined",
  danger:"danger",
  dangerOutlined:"danger-outlined",
  error:"error",
  errorOutlined:"error-outlined",
  default: "default",
  defaultOutlined: "default-outlined",
  disabled: "disabled",
}