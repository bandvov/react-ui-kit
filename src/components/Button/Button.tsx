import React, { FC, ReactElement } from "react";
import styled, { css } from "styled-components";
import { COLORS, STATUS } from "../../CONSTANTS";
import { IButton } from "../../types";

const StyledButton = styled.button<IButton>`
  display: flex;
  gap: 10px;
  justify-content: ${(props) => props.justify || "center"};
  border: ${(props) => props.border || "none"};
  ${(props) => {
    if (props.disabled) {
      return css`
        background-color: ${COLORS.lightgrey};
        color: ${props.color || COLORS.gray};
      `;
    } else if (props.variant === STATUS.success) {
      return css`
        background-color: ${COLORS.green};
        color: ${props.color || "white"};
      `;
    } else if (props.variant === STATUS.successOutlined) {
      return css`
        background-color: transparent;
        color: ${props.color || COLORS.green};
        outline: ${props.border || "1px solid " + COLORS.green};
      `;
    } else if (props.variant === STATUS.dangerOutlined) {
      return css`
        background-color: transparent;
        color: ${props.color || COLORS.orange};
        outline: ${props.border || "1px solid " + COLORS.orange};
      `;
    } else if (props.variant === STATUS.errorOutlined) {
      return css`
        background-color: transparent;
        color: ${props.color || COLORS.red};
        outline: ${props.border || "1px solid " + COLORS.red};
      `;
    } else if (props.variant === STATUS.defaultOutlined) {
      return css`
        background-color: transparent;
        color: ${props.color || COLORS.Blue};
        outline: ${props.border || "1px solid " + COLORS.Blue};
      `;
    } else if (props.variant === STATUS.danger) {
      return css`
        background-color: ${COLORS.orange};
        color: ${props.color || "white"};
      `;
    } else if (props.variant === STATUS.error) {
      return css`
        background-color: ${COLORS.red};
        color: ${props.color || "white"};
      `;
    } else {
      return css`
        background-color: ${COLORS.Blue};
        color: ${props.color || "white"};
      `;
    }
  }};

  border-radius: ${(props) =>
    props.rounded
      ? "50%"
      : props.borderRadius && !props.rounded
      ? props.borderRadius
      : "5px"};

  padding: ${(props) =>
    props.padding
      ? props.padding
      : props.rounded
      ? ".5rem .7rem"
      : ".5rem 1rem"};
  font-size: ${(props) => (props.fontSize ? props.fontSize : "16px")};
  width: ${(props) => (props.fullWidth ? "100%" : "max-content")};
  background-color: ${(props) => props.backgroundColor || ""};
  :hover {
    cursor: ${(props) => (props.disabled ? "not-allowed" : "pointer")};
  }
  :focus {
    ${(props) => {
      if (
        props.variant === STATUS.successOutlined ||
        props.variant === STATUS.dangerOutlined ||
        props.variant === STATUS.errorOutlined ||
        props.variant === STATUS.defaultOutlined
      ) {
        return css`
          outline: 2px solid black;
        `;
      }
    }}
  }
`;

const Button: FC<IButton> = ({
  children,
  icon,
  iconLeft,
  type = "button",
  ...props
}): ReactElement => {
  return (
    <StyledButton {...props} type={type}>
      {iconLeft}
      {children}
      {icon}
    </StyledButton>
  );
};

export default Button;
