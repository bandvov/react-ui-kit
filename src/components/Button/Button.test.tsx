import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import Button from "./Button";
// Every component that is returned maps 1:1 with the stories, but they already contain all decorators from story level, meta level and global level.
import { COLORS } from "../../CONSTANTS";

// Returns a component that already contain all decorators from story level, meta level and global level.

describe("Button tests", () => {
  test("Test default button styles", () => {
    render(<Button onClick={() => {}}>Default</Button>);
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      backgroundColor: COLORS.Blue,
      fontSize: "16px",
      padding: ".5rem 1rem",
      width: "max-content",
      borderRadius: "5px",
      color: "white",
    });
  });
  test("Test Default outlined button styles", () => {
    render(
      <Button variant="default-outlined" onClick={() => {}}>
        Default outlined
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      backgroundColor: "transparent",
      color: COLORS.Blue,
      outline: `1px solid ${COLORS.Blue}`,
      padding: ".5rem 1rem",
      // border-radius: 5px;
    });
  });
  test("Test Success button styles", () => {
    render(
      <Button variant="success" onClick={() => {}}>
        Success
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      backgroundColor: COLORS.green,
    });
  });
  test("Test Success outlined button styles", () => {
    render(
      <Button variant="success-outlined" onClick={() => {}}>
        Success outlined
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      backgroundColor: "transparent",
      color: COLORS.green,
      outline: `1px solid ${COLORS.green}`,
      padding: ".5rem 1rem",
      // border-radius: 5px;
    });
  });
  test("Test Danger button styles", () => {
    render(
      <Button variant="danger" onClick={() => {}}>
        Danger
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      backgroundColor: COLORS.orange,
      padding: ".5rem 1rem",
      // border-radius: 5px;
    });
  });
  test("Test Danger outlined button styles", () => {
    render(
      <Button variant="danger-outlined" onClick={() => {}}>
        Danger outlined
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      backgroundColor: "transparent",
      color: COLORS.orange,
      outline: `1px solid ${COLORS.orange}`,
      padding: ".5rem 1rem",
      // border-radius: 5px;
    });
  });
  test("Test Error button styles", () => {
    render(
      <Button variant="error" onClick={() => {}}>
        Error
      </Button>
    );
    const buttonElement = screen.getByText(/Error/i);
    expect(buttonElement).toHaveStyle({
      backgroundColor: COLORS.red,
      padding: ".5rem 1rem",
      // border-radius: 5px;
    });
  });
  test("Test Error outlined button styles", () => {
    render(
      <Button variant="error-outlined" onClick={() => {}}>
        Error outlined
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      backgroundColor: "transparent",
      color: COLORS.red,
      outline: `1px solid ${COLORS.red}`,
      padding: ".5rem 1rem",
      // border-radius: 5px;
    });
  });
  test("Test button border radius", () => {
    render(
      <Button borderRadius="15px" disabled onClick={() => {}}>
        radius
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      borderRadius: "15px",
    });
  });

  test("Test rounded button styles", () => {
    render(
      <Button rounded onClick={() => {}}>
        rounded
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      borderRadius: "50%",
    });
  });
  test("Test button font size", () => {
    render(
      <Button fontSize="25px" onClick={() => {}}>
        font
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      fontSize: "25px",
    });
  });
  test("Test button full width", () => {
    render(
      <Button fullWidth onClick={() => {}}>
        fullwidth
      </Button>
    );
    const buttonElement = screen.getByTestId(/button/i);
    expect(buttonElement).toHaveStyle({
      width: "100%",
    });
  });
  test("Button action", () => {
    const handler = jest.fn();
    render(<Button onClick={handler}>Button</Button>);
    const buttonElement = screen.getByText(/Button/i);
    fireEvent.click(buttonElement);
    expect(handler).toHaveBeenCalledTimes(1);
  });
});
