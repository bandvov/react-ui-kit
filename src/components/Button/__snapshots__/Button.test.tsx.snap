// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tests Test rounded button styles 1`] = `
<button
  class="sc-bdvvtL jpKAiS"
  type="button"
>
  rounded
</button>
`;
